### ISSUE TITLE
Image Loading Enhancement Discussion

### ISSUE THREAD

**Issue by slembcke** [cocos2d/cocos2d-objc#473]
## Automatic downscaling of PNG and JPEG assets. (Easy)

I've talked about this a lot before so I'll keep this short. It saves devs an annoying step, and helps optimize download sizes without using extra memory at runtime (or load time). This would be used as an optional fallback. I have a prototype 

Possible downsides include 3-4x slower loading on older devices, will add libpng/libjpeg dependencies, and add a couple hundred lines of extra code. I have a working prototype for PNG and it's ~200 lines of code. JPEG will be easier.

What about .pvr files? Downscaling .pvr files is pretty impractical. It is possible to create .pvr files with mipmaps in them, and we could use the scale factor to skip reading the first couple of mipmap levels. Same overall effect, but without some of the disk savings. This is similar to how most 3D engines support configurable texture resolutions on the client as far as I know.
## Built in encryption for images. (Easy)

People got surprisingly excited about this idea when it first came up. If use the custom loading code from above, this should be very easy. I would propose using Texture Packer's existing encryption, but use a .ccp extension. I asked Andreas Löw about his thoughts on this about a month ago, but haven't heard back from him yet.
## GLKTexture Based Texture Loader (very easy)

Alternatively we could simplify texture loading significantly by using GLKit texture loading instead. Automatic downscaling will definitely not be compatible. GZipped or encrypted files are still possible at the expense of wasting memory. 
## PDF or WebP File Support (easy)

People have wanted vector image importing for a long time. Loading sprites from PDFs should be pretty easy. It would even share a lot of the same code as the current ImageIO based loader. Since CoreGraphics already natively supports PDF, it wouldn't even require additional libraries. What about SVG? I have yet to see any particularly good libraries due to it's complexity. SVGKit looks okay, but is _huge_ and still incomplete.

Webp is a really interesting format. It supports better lossless compression than PNG, and better lossy compression than JPEG and includes alpha channel support. It seems perfect for game graphics. It's biggest problem is that it has nearly zero editor support. We would need to provide some sort of asset conversion support via Xcode build phases to make it practical to use.
## My Thoughts:

Downscaling:
I feel that the only downside of automatic downscaling is adding libjpeg and libpng dependencies. We can downscale on the fly with some simple changes to the existing ImageIO code, but I'm not sure about the quality, memory usage or performance. I suspect all three to be vastly inferior. I feel that the risk for this feature is small.

Encryption:
If we refactor the texture loading routines to take an NSStream object instead of a plain file path, we could very easily support optional encryption. I feel that the risk for this feature is small.

GLKTexture Loader:
I was a proponent of GLKit texture loading in the past, but now I feel like it will require dropping a lot of current minor features and preventing new ones like the auto downscaling. Supporting gzipped or encrypted assets is still possible, but it looks like it would require loading them into memory and then passing the pointer to GLKit. Not terribly efficient.

PDF/WebP:
PDFs might be really easy, though I would consider it lower priority than downscaling or encryption. Might as well improve the regular workflow before adding new formats. WebP's lack of editor support makes it very low priority, though combined with some sort of asset workflow in Xcode it could be really nice.

I personally am quite interested in adding the downscaling and encryption support.

**Comment by billhollings** [comment]
## Non-premultiplied Alpha

Conspicuously missing from your list is a way to avoid the automatic pre-multiplying of alpha in PNG & TGA (TGA as of latest Xcode) textures. 

I'd like to add that for discussion.

Is the option for unmangled PNG & TGA files handled already somewhere within cocos2d texture loading, or is it accepted that all PNG's will and should be "optimized" by Xcode and read as such by CGImage?

Although I fully understand the arguments for using pre-multiplied alpha, there are times in the 3D environment where a pure PNG or TGA is required. The obvious case is when the texture is carrying something like a normal map, lighting map, etc. And I have seen developers use textures to hold their own custom content as well (eg- carry shininess in the alpha channel).

Thoughts?

**Comment by slembcke** [comment]
Ah definitely. I actually handled pre-multiplied alpha explicitly with my experimental PNG loader so that its completely optional. https://github.com/cocos2d/cocos2d-iphone/blob/fancy-png/cocos2d/CCTextureCache.m#L409 I'm a terrible artist so most of the textures I make are special effects lookup tables and such. Pre-multiplying the alpha usually ruins them, so I completely agree.

GLKit supports optional pre-multiplication, though I read something in the docs somewhere that gave me the distinct impression that it loaded the images using ImageIO and then un-multiplied the premultiplied alpha. If true that would still be pretty useless. I've been meaning to make an example to test it out, but haven't gotten around to it. I also can't find the doc that made me think that.

I remember in the past that it _seemed_ like the APIs should let you get non-premultiplied alpha with ImageIO (ex: kCGImageAlphaLast vs. kCGImageAlphaPremultipliedLast), but I remember struggling to get CoreGraphics to accept anything using the non-premultiplied types as being a valid CGBitmapContext format and that is the only way I know to get the bitmap out of a CGImage.

**Comment by billhollings** [comment]
All the research I've done (only trolling documentation so far) leads me to believe that it is not possible in practice to use CoreGraphics to load un-premultiplied PNG's & TGA's (don't know about GLKit yet).

For **cocos3d**, I'm thinking of using [stb_image](http://nothings.org/stb_image.c) as an external library to load PNG's and TGA's that have not been pre-multiplied. It's smaller than the default libpng, etc, and has some good reviews.

The other issue is how to avoid Xcode munging the files in the first place. My intention there is to use specialized suffixes (say - .pngu & .tgau - 'u' being a reference to "UN-pre-multiplied") to fool Xcode, and to indicate how the files should be loaded.

**Comment by slembcke** [comment]
stb_image is sort of neat, but it's missing a ton of features and a lot of files will simply fail to load with it. I only had to deal with that once before deciding that 150 lines of libPNG code wasn't so bad.

If you want to prevent Xcode from mangling your PNGs there is a setting for it. Apple calls it "compressing" instead of "corrupting".
![Xcode Setting](http://files.slembcke.net/upshot/upshot_A9tl0mao.png)

You need to disable that if you want the files to be loadable with anything other than ImageIO on iOS.

**Comment by billhollings** [comment]
Fair enough about libPNG vs stb_image. Either way, I intend only to use them for the specialized loading. Regular PNG's and TGA's would go through the standard CG path.

Two problems with that Xcode setting:
1. No one seems to be able to get it to actually work in practice. If you have success, please let me know what your secret ingredient is!  :^)
2. If it does work at all, it will do so for all PNG files. Since premultiplication is actually useful, it would be best to identify only those files you _don't_ want to premultiply...both to Xcode and the loader. Hence the specialized suffix.

**Comment by slembcke** [comment]
- No magic. I just clicked it and it worked. Maybe you need to do a clean build?
- It does it for all files yes, but it doesn't break anything. AFAIK all it does is change it from RGBA to BGRA order, premultiply, and set some non-standard tags that libPNG chokes on. ImageIO can still read normal non-premultiplied, RGBA PNG files just fine. Otherwise you wouldn't be able to load files from the internet or off the file system. In theory it helped performance on old devices by skipping some conversion steps, but it doesn't seem to make much of a difference anymore.

**Comment by cocojoe** [comment]
PDF/WebP:
I did some tests with WebP ages back: http://www.cocos2d-iphone.org/forums/topic/reduce-image-size-webp-googles-image-format-in-practice/
It's a nice format and the filesize/quality is excellent, shame no one uses it :-) 
Who creates their assets in PDF? Is there really demand for this?

Downscaling:
Auto downscaling is a nice although the actual size of non retina assets is pretty small, would it be easier just to have it done as part of the build rather than runtime? Low resolution well become a non issue in the future.

Encryption:
This is a must have for me.

GLKTexture Loader:
I was pro this as I'm all for using native functionality vs rewriting the wheel however seems like to many restrictions as you mention.

**Comment by slembcke** [comment]
Do people make assets in PDF? No, but Cocos doesn't currently support loading them yet either. ;) I've worked on several projects where the artists used Illustrator and they really wished they could just use vector files in the actual app. Especially for some of the really large background images where the file size was an issue. Plenty of people have asked for vector graphics support on the forums too. SVG might be more popular as a vector format, but it's not easy to support while PDF is. I suspect it will be no more than ~50 lines of code to implement it.

I feel that the downscaling is more of a convenience to make it easier to create and manage assets. If we could make some sort of an asset importing pipeline work well with an Xcode build rule, that would be about as good. Due to the limitations of build rules, I've struggled with that in the past though. Saving 20-25% of your disk space for textures isn't a bad reason to support downscaling either though.
